clients/drcachesim/common/memref.h

```diff
@@ -49,8 +49,71 @@ namespace drmemtrace { /**< DrMemtrace tracing + simulation infrastructure names
 
 // On some platforms, like MacOS, a thread id is 64 bits.
 // We just make both 64 bits to cover all our bases.
-typedef int64_t memref_pid_t; /**< Process id type. */
-typedef int64_t memref_tid_t; /**< Thread id type. */
+/**
+ * Process id type.  When multiple workloads are combined in one trace, a workload ordinal
+ * is added to the top (64-MEMREF_ID_WORKLOAD_SHIFT) bits; the workload_from_memref_pid()
+ * and pid_from_memref_pid() helpers can be used to separate the values if desired.
+ */
+typedef int64_t memref_pid_t;
+/**
+ * Thread id type.  When multiple workloads are combined in one trace, a workload ordinal
+ * is added to the top (64-MEMREF_ID_WORKLOAD_SHIFT) bits; the workload_from_memref_tid()
+ * and tid_from_memref_tid() helpers can be used to separate the values if desired.
+ */
+typedef int64_t memref_tid_t;
+
+/** Constants related to tid and pid fields. */
+enum {
+    /**
+     * When multiple workloads are combined in one trace, a workload ordinal is added to
+     * the top (64-MEMREF_ID_WORKLOAD_SHIFT) bits of the pid and tid fields of #memref_t.
+     */
+    MEMREF_ID_WORKLOAD_SHIFT = 48,
+};
+
+/**
+ * When multiple workloads are combined in one trace, a workload ordinal is added to the
+ * top (64-MEMREF_ID_WORKLOAD_SHIFT) bits of the #memref_t pid field.  This function
+ * extracts the workload ordinal.
+ */
+static inline int
+workload_from_memref_pid(memref_tid_t pid)
```

Comment:
We likely want new functions to be in the dynamorio::drmemtrace namespace?

Reply:
These are inside that namespace: the namespace is above on lines 47-48.